controllers/fenceagentsremediation_controller.go

```diff
@@ -84,6 +85,29 @@ func (r *FenceAgentsRemediationReconciler) Reconcile(ctx context.Context, req ct
 		r.Log.Error(err, "failed to get FAR CR")
 		return emptyResult, err
 	}
+	// Add finalizer when object is created
+	if far.ObjectMeta.DeletionTimestamp.IsZero() {
+		controllerutil.AddFinalizer(far, v1alpha1.FARFinalizer)
+		if err := r.Client.Update(context.Background(), far); err != nil {
+			return emptyResult, fmt.Errorf("failed to add finalizer to the CR - %w", err)
+		}
+	} else if controllerutil.ContainsFinalizer(far, v1alpha1.FARFinalizer) {
+		r.Log.Info("CR's deletion timestamp is not zero, and FAR finalizer exists", "CR Name", req.Name)
+		// The object is being deleted
+
+		// remove node's taints
+		if err := utils.RemoveTaint(r.Client, far.Name); err != nil && !apiErrors.IsNotFound(err) {
+			return emptyResult, err
+		}
+		// remove finalizer
+		controllerutil.RemoveFinalizer(far, v1alpha1.FARFinalizer)
+		if err := r.Client.Update(context.Background(), far); err != nil {
+			return emptyResult, fmt.Errorf("failed to remove finalizer from CR - %w", err)
+		}
+		r.Log.Info("Finalizer was removed", "CR Name", req.Name)
+		return emptyResult, nil
```

Comment:
you need to handle the case with deletionTimestamp set, but no finalizer set

Reply:
After our discussion I see that I have neglected the scenario of `when the CR is deleted already before you reconcile for the first time` which can happen due to a deletion in a possible delay (e.g., network issues, node issues, and etc.).